src/value_serializer.rs

```diff
@@ -0,0 +1,109 @@
+use crate::Context;
+use crate::HandleScope;
+use crate::Isolate;
+use crate::Local;
+
+use crate::support::MaybeBool;
+
+use crate::data::ArrayBuffer;
+use crate::data::Value;
+
+use std::ffi::c_void;
+use std::mem::MaybeUninit;
+
+/// Instances of this class can be passed
+// Must be >= sizeof(v8::ValueSerializer), see v8__ValueSerializer__CONSTRUCT().
+#[repr(C)]
+pub struct ValueSerializer([usize; 1]);
+
+extern "C" {
+  fn v8__ValueSerializer__CONSTRUCT(
+    buf: *mut MaybeUninit<ValueSerializer>,
+    isolate: *mut Isolate,
+  );
+  fn v8__ValueSerializer__DESTRUCT(this: *mut ValueSerializer);
+
+  fn v8__ValueSerializer__TransferArrayBuffer(
+    this: *mut ValueSerializer,
+    transfer_id: u32,
+    array_buffer: Local<ArrayBuffer>,
+  );
+
+  fn v8__ValueSerializer__WriteHeader(this: *mut ValueSerializer);
+  fn v8__ValueSerializer__WriteValue(
+    this: *mut ValueSerializer,
+    context: Local<Context>,
+    value: Local<Value>,
+  ) -> MaybeBool;
+  fn v8__ValueSerializer__WriteUint32(this: *mut ValueSerializer, value: u32);
+  fn v8__ValueSerializer__WriteUint64(this: *mut ValueSerializer, value: u64);
+  fn v8__ValueSerializer__WriteDouble(this: *mut ValueSerializer, value: f64);
+  fn v8__ValueSerializer__WriteRawBytes(
+    this: *mut ValueSerializer,
+    source: *const c_void,
+    length: usize,
+  );
+}
+
+impl ValueSerializer {
+  pub fn new<'s>(scope: &mut HandleScope<'s>) -> Self {
+    let isolate: *mut Isolate = scope.get_isolate_ptr();
+    let mut value_serializer: MaybeUninit<Self> = MaybeUninit::uninit();
+    unsafe {
+      v8__ValueSerializer__CONSTRUCT(&mut value_serializer, isolate);
+      value_serializer.assume_init()
+    }
+  }
+}
+
+impl Drop for ValueSerializer {
+  fn drop(&mut self) {
+    unsafe { v8__ValueSerializer__DESTRUCT(self) };
+  }
+}
+
+impl ValueSerializer {
+  pub fn transfer_array_buffer(
+    &mut self,
+    transfer_id: u32,
+    array_buffer: Local<ArrayBuffer>,
+  ) {
+    unsafe {
+      v8__ValueSerializer__TransferArrayBuffer(self, transfer_id, array_buffer)
+    };
+  }
+}
+
+impl ValueSerializer {
+  pub fn write_header(&mut self) {
+    unsafe { v8__ValueSerializer__WriteHeader(self) };
+  }
+
+  pub fn write_value(
+    &mut self,
+    context: Local<Context>,
+    value: Local<Value>,
+  ) -> Option<bool> {
+    unsafe { v8__ValueSerializer__WriteValue(self, context, value) }.into()
+  }
+
+  pub fn write_uint32(&mut self, value: u32) {
+    unsafe { v8__ValueSerializer__WriteUint32(self, value) };
+  }
+
+  pub fn write_uint64(&mut self, value: u64) {
+    unsafe { v8__ValueSerializer__WriteUint64(self, value) };
+  }
+
+  pub fn write_double(&mut self, value: f64) {
+    unsafe { v8__ValueSerializer__WriteDouble(self, value) };
+  }
+
+  pub unsafe fn write_raw_bytes(
```

Comment:
We try to adapt function/method signatures in such way that they can be used without `unsafe` and do not use raw pointers.

I think you could make `write_raw_bytes()` safe by doing something along these lines:

```rust
pub fn write_raw_bytes(
    &mut self,
    source: &[u8]
  ) {
    unsafe { v8__ValueSerializer__WriteRawBytes(self, source.as_ptr() as *const _, source.len()); }
  }
}
```